src/protocol/mod.rs

```diff
@@ -295,6 +296,18 @@ impl Protocol {
     }
 
     fn from_object(object: &Hash) -> R<Protocol> {
+        check_keys(
```

Comment:
I think it's definitely possible. I'm a bit cautious that this might be over-engineering. I agree that ideally we would somehow tie the querying of the keys to the check for unknown keys, so that

a) developers can't query a key that will be flagged as unknown and
b) developers can't allow a key that is not used.

As long as we stick to TDD a) shouldn't be a problem, since the tests for reading a new field would not pass. b) can happen, for example if we rename a field and a developer accidentally adds the renamed version, without removing the old one. Doesn't seem very likely though.

So while I think this sounds like a fun thing to think about, I don't think it's *that* important.